src/qutip_qip/qubits.py

```diff
@@ -31,36 +31,229 @@
 
-from qutip import (tensor, basis)
+import qutip
+import numpy as np
 from numpy import sqrt
 
 
+def qubit_states(states):
+    """
+    Shortcut to generate disentangled qubit states.
+
+    Parameters
+    ----------
+    states : list or str
+        - If a list consisting of ``0``, ``1``, ``"0"``, ``"1"``, ``"+"``
+          and ``"-"``, return the corresponding zero/one/plus/minus state.
+        - If a string consisting of ``0``, ``1``, ``+``, ``-``,
+          same as above.
+        - If a list of float or complex numbers,
+          each number is mapped to a state of the form
+          :math:`\\sqrt{1 - |a|^2} \\left|0\\right\\rangle + a |1\\rangle`,
+          where :math:`a` is the given number.
+
+    Returns
+    -------
+    quantum_states : :obj:`qutip.Qobj`
+        The generated qubit states.
+
+    Examples
+    --------
+    >>> from qutip_qip.qubits import qubit_states
+    >>> qubit_states([0, 0])  # doctest: +NORMALIZE_WHITESPACE
+    Quantum object: dims = [[2, 2], [1, 1]], shape = (4, 1), type = ket
+    Qobj data =
+    [[1.]
+     [0.]
+     [0.]
+     [0.]]
+    >>> qubit_states([1, "+"])  # doctest: +NORMALIZE_WHITESPACE
+    Quantum object: dims = [[2, 2], [1, 1]], shape =
+    (4, 1), type = ket
+    Qobj data =
+    [[0.        ]
+     [0.        ]
+     [0.70710678]
+     [0.70710678]]
+    >>> qubit_states("-")  # doctest: +NORMALIZE_WHITESPACE
+    Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
+    Qobj data =
+    [[ 0.70710678]
+     [-0.70710678]]
+    >>> qubit_states("1-")  # doctest: +NORMALIZE_WHITESPACE
+    Quantum object: dims = [[2, 2], [1, 1]], shape =
+    (4, 1), type = ket
+    Qobj data =
+    [[ 0.        ]
+     [ 0.        ]
+     [ 0.70710678]
+     [-0.70710678]]
+    >>> import numpy as np
+    >>> qubit_states([1.j/np.sqrt(2)])  # doctest: +NORMALIZE_WHITESPACE
+    Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
+    Qobj data =
+    [[0.70710678+0.j        ]
+     [0.        +0.70710678j]]
+    """
+    states_map = {
+        0: qutip.basis(2, 0),
+        1: qutip.basis(2, 1),
+        "0": qutip.basis(2, 0),
+        "1": qutip.basis(2, 1),
+        "+": (qutip.basis(2, 0) + qutip.basis(2, 1)).unit(),
+        "-": (qutip.basis(2, 0) - qutip.basis(2, 1)).unit(),
+    }
+
+    states_list = []
+    for s in states:
+        if s in states_map:
+            states_list.append(states_map[s])
+        elif np.isscalar(s) and abs(s) <= 1:
+            states_list.append(
+                s * qutip.basis(2, 1) +
+                np.sqrt(1 - abs(s)**2) * qutip.basis(2, 0)
+            )
+        else:
+            raise TypeError(f"Invalid input {s}.")
+    return qutip.tensor(states_list)
+
+
+def _find_reduced_indices(dims):
+    """Find the forresponding indices for a given qu"""
+    if len(dims) == 1:
+        return np.array([0, 1], dtype=np.intp)
+    else:
+        rest_reduced_indices = _find_reduced_indices(dims[1:])
```

Comment:
@BoxiLi I might be unfamiliar with how you defined the function here. 

How can you use a function being defined in its own definition ?

Reply:
Good question, this is called "recursion". A function calls itself indeed but every time the problem gets simpler. And there is always a "leaf" that stops this (the first `if` branch), otherwise, it will run forever.  In this code, every time I call the function, the dims becomes shorter. and eventually, it becomes length 1 and the first `if` branch terminates the program.

Reply:
The best example is always the Fibonacci numbers in Python that you can find everywhere.

``` python
def fibonacci(n):
    elif n == 0:
        return 0
    elif n == 1 or n == 2:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)
```

Reply:
That's pretty interesting ! I knew about recursion of variables but this seems more efficient ? Here, you could skip defining the function variables for recursion, define a loop for them etc. So, is there a name for this method or it is another possible form of recursion ?

It's interesting that you can use `elif` without an `if` as well. I always thought the order was `if --> elif --> else` even in function definition. But since your function will first try to run the code in `else` loop, `elif` portions will be run after `return fibonacci(n-1) + fibonacci(n-2)`.

Reply:
> It's interesting that you can use elif without an if as well. 
Ohhhh nonono, that's not true. that was just a bug. You should always start with an `if`.

Reply:
> Ohhhh nonono, that's not true. that was just a bug

Ah. Ok. I haven't tried the code yet but my comment was looking through the code itself.

Reply:
> I knew about recursion of variables but this seems more efficient? Here, you could skip defining the function variables for recursion, define a loop for them etc. So, is there a name for this method or it is another possible form of recursion ?

I'm not sure if I understand correctly what you mean by *recursion of variables*. But in my use case here, you can indeed replace the recursion with a `for` loop. Because you know that it will take `len(dims)` iterations. However, that is not always true, like e.g. recursive calculation of the [Collatz conjecture](https://en.wikipedia.org/wiki/Collatz_conjecture).

Here, I choose the recursion instead of a `for` loop, because if I wrote a loop, I would have to start from the last item in `dims` (the last qubit). The nice thing about recursion is that you can understand the line
```rest_reduced_indices = _find_reduced_indices(dims[1:])```
as 
>Assuming that I know the result of `_find_reduced_indices(dims[1:])`, I only need to deal with the first indices. Then I can do it in the following way..."

The recursion helps you "realize" that assumption, which makes the logic cleaner.